packages/react/src/ReactFetch.js

```diff
@@ -0,0 +1,130 @@
+/**
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ *
+ * @flow
+ */
+
+import {
+  enableCache,
+  enableFetchInstrumentation,
+} from 'shared/ReactFeatureFlags';
+
+import ReactCurrentCache from './ReactCurrentCache';
+
+function createFetchCache(): Map<string, Array<any>> {
+  return new Map();
+}
+
+const simpleCacheKey = '["GET",[],null,"follow",null,null,null,null]'; // generateCacheKey(new Request('https://blank'));
+
+function generateCacheKey(request: Request): string {
+  // We pick the fields that goes into the key used to dedupe requests.
+  // We don't include the `cache` field, because we end up using whatever
+  // caching resulted from the first request.
+  // Notably we currently don't consider non-standard (or future) options.
+  // This might not be safe. TODO: warn for non-standard extensions differing.
+  // IF YOU CHANGE THIS UPDATE THE simpleCacheKey ABOVE.
+  return JSON.stringify([
+    request.method,
+    Array.from(request.headers.entries()),
+    request.mode,
+    request.redirect,
+    request.credentials,
+    request.referrer,
+    request.referrerPolicy,
+    request.integrity,
+  ]);
+}
+
+if (enableCache && enableFetchInstrumentation) {
+  if (typeof fetch === 'function') {
+    const originalFetch = fetch;
+    try {
+      // eslint-disable-next-line no-native-reassign
+      fetch = function fetch(
+        resource: URL | RequestInfo,
+        options?: RequestOptions,
+      ) {
+        const dispatcher = ReactCurrentCache.current;
+        if (!dispatcher) {
+          // We're outside a cached scope.
+          return originalFetch(resource, options);
+        }
+        if (
+          options &&
+          options.signal &&
+          options.signal !== dispatcher.getCacheSignal()
+        ) {
+          // If we're passed a signal that is not ours, then we assume that
+          // someone else controls the lifetime of this object and opts out of
+          // caching. It's effectively the opt-out mechanism.
+          // Ideally we should be able to check this on the Request but
+          // it always gets initialized with its own signal so we don't
+          // know if it's supposed to override - unless we also override the
+          // Request constructor.
+          return originalFetch(resource, options);
+        }
+        // Normalize the Request
+        let url: string;
```

Comment:
Do you need to normalize url if it's a relative path string? In case history.push gets called while the cache is alive.

Reply:
Only for the fast path. Good catch. Originally I didn't have the fast path.